src/org/sosy_lab/java_smt/solvers/mathsat5/Mathsat5AbstractProver.java

```diff
@@ -100,21 +98,33 @@ private long buildConfig(Set<ProverOptions> opts) {
   public boolean isUnsat() throws InterruptedException, SolverException {
     Preconditions.checkState(!closed);
 
-    final long hook = msat_set_termination_callback(curEnv, context.getTerminationTest());
+    final long hook = msat_set_termination_callback(curEnv, getTerminationTest());
     try {
       return !msat_check_sat(curEnv);
     } finally {
       msat_free_termination_callback(hook);
     }
   }
 
+  /**
+   * Get a termination callback for the current prover (who's parent is the contexts callback). The
+   * callback can be registered upfront, i.e., before calling a possibly expensive computation in
+   * the solver to allow a proper shutdown.
+   */
+  private TerminationCallback getTerminationTest() {
+    return () -> {
+      proverShutdownManager.getNotifier().shutdownIfNecessary();
```

Comment:
We could extract the reference to the notifier and just reuse.

Might look like this:


```
final ShutdownNotifier notifier = getShutdownManagerForProver().getNotifier()
return () -> {
  notifier.shutdownIfNecessary();
  return false;
};
```

Reply:
Done! Thanks for the hint.